### ISSUE TITLE
Multiple instances of SDLRouter running confuses SDLProxy, not connecting to the router with a HU connection.

### ISSUE THREAD

**Issue by jthrun** [smartdevicelink/sdl_android#307]
In the generic situation this is probably an edge case where two SDL apps check for a running router almost simultaneously and, finding none, each start their routers resulting in two routers running at the same time.  Our situation is different in that we have a companion app that only starts its router once it detects a connection to a known head unit.  The companion app's router will be started regardless of other routers running since only the companion app's router will get a connection from the HU SDLCore.  Unfortunately, the other SDL apps do not always find and connect to this router even though it broadcasts the 'sdl.router.startservice' message for triggering SDLProxy startup.  What I see in the log is that SdlBroadcastReceiver validates the router that broadcast the message but the proxy (or actually the TransportBroker) later tries to check for a transport connected router and always finds the non-connected router first, short-circuiting and returning false.

In order to repro this you need a "companion" app that will start its SdlRouter when it gets a connection to a "known" HU.  This app's SdlRouter needs to use a custom BT UUID which the SDLCore on the "known" HU will connect to to the exclusion of other SdlRouters broadcasting with the standard UUID.

1. Install the "companion" app and other SDL multiplexer-enabled apps on the HS device
2. Restart the HS
3. Start the HU, SDLCore, HMI
4. Try to run one of the SDL apps

What I see in the logs is that the "companion" app's router gets a connection from SDLCore on the HU and then broadcasts the TransportConstants.START_ROUTER_SERVICE_ACTION message.  Then, other apps' SdlBroadcastReciever check and validate the broadcasting router.  Finally, SDLProxies for those apps check to see of a running router has a transport connection but get back a false response because they checked the wrong running router.  I'm attaching a section of logcat logs which contains messages from two routers (pid 3504, 5368) and one of the apps (pid 3630) which shows what happens.  Short story:
```
06-21 13:30:26.974  3630  3630 V Sdl Broadcast Receiver: SDL enabled by router service from com.uievolution.companionapp compnent package com.uievolution.companionapp - com.main.companionapp.sdl.SdlRouterService
...
06-21 13:30:27.028  3630  3630 D SdlTransportBroker: Bound to service ComponentInfo{com.uievolution.companionapp/com.main.companionapp.sdl.SdlRouterService}
...
06-21 13:30:27.031  3630  5561 D Sdl Broadcast Receiver: Looking for Service: sdlrouterservice
06-21 13:30:27.034  3630  5561 D Sdl Broadcast Receiver: Found service: com.uievolution.sdl.CTSMediaApp1.livio.SdlRouterService in package com.uievolution.sdl.CTSMediaApp1
06-21 13:30:27.038  3630  5561 W Sdl Broadcast Receiver: Is Connected? Returning false

```
Note in the first two lines the app is binding to the correct router but in the last three lines it checks the connection state of the incorrect router.

I have thought of a couple ways of dealing with this issue.  One way would be to have each app shut down its own router if the router is not connected to the HU.  Another option would be to propagate the package/service information about the connected router from the broadcast receiver to all parts of the code that need to check the router's connection.  I'll create a pull request for a solution that checks the broadcasting router to see if it is transport connected.

**Comment by jthrun** [comment]
[issue.txt](https://github.com/smartdevicelink/sdl_android/files/326613/issue.txt)

**Comment by joeygrover** [comment]
So the issue stems from having a router service that does not obey the current "correct" lifecycle flow. The companion app will always start up and keep up its router service regardless of whether or not another router service is running that has went through the correct lifecycle flow. If the routers obey the correct flow this issue doesn't propagate, correct?

**Comment by jthrun** [comment]
Yes, mostly, I guess.  Though I have seen a couple times where the race-condition case, that I mentioned at the top, occured such that there were two, unconnected, generic UUID routers running at the same time.  I assumed this occured in the edge condition of two SDL apps checking almost simultaneously for a router and, neither finding one, both starting their own.  Also, with the current "correct" flow there doesn't seem to be a "correct" way for different routers to use different UUIDs to advertise their services, which is what we're trying to do.

**Comment by joeygrover** [comment]
Hmm I've never seen it with the normal flow, but that's not to say it's impossible to happen. 

Yes, that is by design. In 99% of integrations, no app should be trying to change away from the designated SDL UUID. The correct flow is for one of the already installed SDL enabled apps to propagate the router service preventing the user from having to download an extra app.

**Comment by jthrun** [comment]
Any idea when PR https://github.com/smartdevicelink/sdl_android/pull/312 will be reviewed/merged?

**Comment by joeygrover** [comment]
@mrapitis any update on the review?

**Comment by joeygrover** [comment]
This should be resolved with PR #312 being merged into develop. @jthrun can you confirm?

**Comment by jthrun** [comment]
This issue is fixed.  I'm not seeing the initial connection issue that was happening before.
